added flyio detector

lonmarsDev opened the pull request:
<!--
Please create an issue to collect feedback prior to feature additions. Please also reference that issue in any PRs.
If possible try to keep PRs scoped to one feature, and add tests for new features.
-->

### Description:
Explain the purpose of the PR.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

lonmarsDev committed:
added flyio detector

lonmarsDev committed:
Merge branch 'main' into detectors/flyio_detector

zricethezav commented on pkg/detectors/flyio/flyio.go:
```diff
@@ -0,0 +1,104 @@
+var (
+	defaultClient = common.SaneHttpClient()
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"flyio"}) + `\b([a-zA-Z0-9_]{47})\b`)
```
@lonmarsDev all the tokens I generated started with `fo1_`. Does that match what you were seeing? If so we should probably add `fo1_` as part of this regex and to the keywords.

btoews replied:
Fly.io has a few legacy token formats that are no longer issued, but still might exist in the wild:

- `[a-zA-Z0-9_-]{43}`
- `fm1[ar]_[a-zA-Z0-9+\/]={0,2}`

And two formats that we currently issue

- `fo1_[a-zA-Z0-9_-]{43}`
- `fm2_[a-zA-Z0-9+\/]={0,2}`

amanfcp replied:
@btoews Thanks. Can you please give me a resource/link where this is mentioned? Asking as I need to document it.

amanfcp replied:
So far, the tokens (both app and org tokens) that I have created start with `FlyV1 fm2_`. Haven't seen a token with `fo1_` prefix yet.

zricethezav commented:
We might want to rename this as `flyiopersonalaccesstoken` since fly.io also has organization tokens (see https://fly.io/docs/flyctl/tokens-create-org/ and `https://fly.io/dashboard/{user_id}/tokens`). Alternatively we could just leave this as `flyio` and use `flyioorgtoken` for the organizational token detector when we get to that.

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' of github.com:trufflesecurity/trufflehog into detectors/flyio_detector

amanfcp committed:
Add Fly.io Detector with Integration Tests

- Adjusted regex patterns for token matching
- Introduced a new Fly.io secret detector with functionality to identify and verify Fly.io tokens.
- Implemented integration tests to validate the detection and verification logic, including handling of false positives.
- Updated existing Fly.io test cases to enhance coverage and accuracy.
- Added logic to bypass common false positive checks for valid tokens.

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

nabeelalam commented on pkg/detectors/flyio/flyio.go:
```diff
@@ -0,0 +1,150 @@
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"flyio", "FlyV1", "fm2_"}
```
`fm2_` seems like a commonly occurring substring so I'm not sure if it would be an effective keyword. I think `FlyV1` is specific enough to be the sole required keyword here, since it's always part of the secret.

nabeelalam commented on pkg/detectors/flyio/flyio.go:
```diff
@@ -0,0 +1,150 @@
+	// 403 is returned if incorrect org_slug is sent.
+	// 401 is returned if the token is invalid.
+	// 400 is returned if the token is valid but no org_slug is sent.
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.machines.dev/v1/apps?org_slug=", nil)
```
nit: Can use `http.NoBody` instead of `nil`

amanfcp committed:
incorporated feedback on PR

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

kashifkhan0771 committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

shahzadhaider1 committed:
Merge branch 'main' into detectors/flyio_detector

shahzadhaider1 commented on pkg/detectors/flyio/flyio.go:
```diff
@@ -0,0 +1,150 @@
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"flyio", "FlyV1"}
```
I noticed that as well, but then I figured that since "flyio" is a relatively unique term, it’s unlikely to introduce noise, so I decided to leave it as is.

amanfcp committed:
Remove "flyio" keyword from Fly.io detector keywords for improved accuracy in secret detection.

amanfcp committed:
Merge branch 'detectors/flyio_detector' of https://github.com/lonmarsdev/trufflehog into detectors/flyio_detector

nabeelalam reviewed (approved):
Thanks for the update @amanfcp !

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
use existing false positive filtering flow while excluding AAAAAA which seems to be present in valid tokens

amanfcp committed:
Merge branch 'detectors/flyio_detector' of https://github.com/lonmarsdev/trufflehog into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

amanfcp committed:
Merge branch 'main' into detectors/flyio_detector

kashifkhan0771 committed:
Merge branch 'main' into detectors/flyio_detector
